File: ruby/word-count/word_count.rb
# frozen_string_literal: true

class Phrase
  def initialize(phrase)
    @phrase = phrase
  end

  attr_reader :phrase

  def word_count
    phrase.downcase.scan(/\b[\w']+\b/)
          .each_with_object(Hash.new(0)) { |word, counter| counter[word] += 1 }
  end
end

File: java/word-count/WordCount.java
import java.util.HashMap;
import java.util.Map;
import java.util.Scanner;

public class WordCount {
    Map<String, Integer> phrase(String phrase) {
        HashMap<String, Integer> counter = new HashMap<>();
        Scanner scanner = new Scanner(phrase.toLowerCase());

        scanner.useDelimiter("[\\s,!@$#%%^&:\\.]+");

        while (scanner.hasNext()) {
            String word = scanner.next();
            char beginning = word.charAt(0);
            if (beginning == word.charAt(word.length() - 1) &&
                    (beginning == '"' || beginning == '\'')) {
                word = word.substring(1, word.length() - 1);
            }
            counter.merge(word, 1, Integer::sum);
        }


        return counter;
    }
}
